stream-chat-android-ui-common/src/main/kotlin/io/getstream/chat/android/ui/common/feature/messages/list/MessagePositionHandler.kt

```diff
@@ -56,31 +59,42 @@ public fun interface MessagePositionHandler {
         @Suppress("ComplexCondition")
         public fun defaultHandler(): MessagePositionHandler {
             return MessagePositionHandler {
-                    previousMessage: Message?,
-                    message: Message,
+                    prevMessage: Message?,
+                    curMessage: Message,
                     nextMessage: Message?,
                     isAfterDateSeparator: Boolean,
+                    isBeforeDateSeparator: Boolean,
                     _: Boolean,
                 ->
-                val previousUser = previousMessage?.user
-                val user = message.user
-                val nextUser = nextMessage?.user
 
+                val curUser = curMessage.user
                 mutableListOf<MessagePosition>().apply {
-                    if (previousMessage == null || previousUser != user || previousMessage.isSystem() ||
-                        isAfterDateSeparator
-                    ) {
-                        add(MessagePosition.TOP)
-                    }
-                    if (previousMessage != null && nextMessage != null && previousUser == user && nextUser == user) {
-                        add(MessagePosition.MIDDLE)
-                    }
-                    if (nextMessage == null || nextUser != user || nextMessage.isSystem() || nextMessage.isError()) {
-                        add(MessagePosition.BOTTOM)
-                    }
+                    val isTop = !isBeforeDateSeparator &&
+                        (isAfterDateSeparator || prevMessage.isInvalidMessageFromUser(curUser)) &&
+                        nextMessage.isValidMessageFromUser(curUser)
+
+                    val isMiddle = !isAfterDateSeparator && !isBeforeDateSeparator &&
+                        prevMessage.isValidMessageFromUser(curUser) &&
+                        nextMessage.isValidMessageFromUser(curUser)
+
+                    val isBottom = !isAfterDateSeparator &&
+                        (isBeforeDateSeparator || nextMessage.isInvalidMessageFromUser(curUser)) &&
+                        prevMessage.isValidMessageFromUser(curUser)
+
+                    if (isTop) add(MessagePosition.TOP)
+                    if (isMiddle) add(MessagePosition.MIDDLE)
+                    if (isBottom) add(MessagePosition.BOTTOM)
                     if (isEmpty()) add(MessagePosition.NONE)
                 }
```

Comment:
Fixed `MessagePosition` determination so that `NONE` represents the case when the message is a single isolated message, and addressed other corner cases.